### DISCUSSION TITLE
PayGreenJS - Évènement de refus de 3DS

### DISCUSSION THREAD

**Discussion by francoistomasi** [PayGreen/documentation#8] [Q&A] [answered]
Bonjour,

Nous souhaitons gérer l'erreur provoquée lors d'un refus de 3DS lors d'une tentative d'enregistrement d'instrument.

```
paygreenjs.init({
    mode: 'instrument',
    modeOptions: {
      authorizedInstrument: true,
    },
    style,
    paymentMethod: 'bank_card',
    publicKey: 'pk_XXXXXXXXX',
  });
```

Or lors du refus, aucun évènement n'est déclenché pas même l'event `paygreenjs.Events.ERROR`

```
{"type":"https:\/\/tools.ietf.org\/html\/rfc2616#section-10","title":"An error occurred","status":500,"detail":"Internal Server Error"}
```

Avec-vous plus d'informations pour ce scénario ?

Merci d'avance,

**Comment by enzobillisPG** [comment]
Bonjour Francois,

L'event `paygreenjs.Events.ERROR` n'est utilisé que pour les erreurs internes à PGJS (une erreur à l'initialisation de PGJS par exemple). 

En mode `instrument` trois possibilités pour gérer les erreurs :

- L'erreur survient à la tokenization, vous pouvez alors être avertis grâce à l'event [TOKEN_FAIL](https://pgjs.paygreen.fr/latest/documentation/enums/types.EventsType.html#TOKEN_FAIL)
- L'erreur survient à la création de l'instrument avant le 3DS, vous recevrez alors un instrument avec un attribut `instrument.authentication.status = failed` en souscrivant à l 'event [INSTRUMENT_READY](https://pgjs.paygreen.fr/latest/documentation/enums/types.EventsType.html#INSTRUMENT_READY)
- L'erreur survient pendant le processus de 3DS, vous recevrez alors un instrument avec un attribut `instrument.authorized = false` en souscrivant à l 'event [INSTRUMENT_READY](https://pgjs.paygreen.fr/latest/documentation/enums/types.EventsType.html#INSTRUMENT_READY)

Bonne journée,
Enzo

**Comment by francoistomasi** [comment]
Merci pour votre retour.

Les 2 évènements `TOKEN_FAIL` et `INSTRUMENT_READY` sont bien déclenchés lors de nos tests avec ces 4 cartes.

PAN | Scheme | Auth method | Response | Explanation
-- | -- | -- | -- | --
4970105715165150 | Visa | 3DSV2 | Success / Refused* | expired_card
4970109000034568 | Visa | 3DSV2 | Success / Refused* | lost_card
4970106155797460 | Visa | 3DSV2 | Success / Refused* | stolen_card
5549650000000124 | Mastercard | 3DSV2 | Success / Refused* | suspected_fraud

Mais lors de nos tests avec la carte Visa valide :

PAN | Scheme | Auth method | Response | Explanation
-- | -- | -- | -- | --
4970105191919435 | Visa | Frictionless | Success |  

Aucun des 2 évènements n'est déclenché et une erreur serveur est renvoyée :
```
{"type":"https:\/\/tools.ietf.org\/html\/rfc2616#section-10","title":"An error occurred","status":500,"detail":"Internal Server Error"}
```

Pouvez-vous vérifier ce comportement ?

Merci d'avance,

**Reply by enzobillisPG** [reply]
Merci pour votre retour, je remonte l'info à notre équipe back et reviens vers vous.
Pouvez-vous me dire si c'est à la tokenization ou à l'autorisation que l'erreur survient ?

**Reply by francoistomasi** [reply]
Merci à vous.
Pour reproduire ce scénario:
- Nous saisissons les infos de la carte suivante:
    - PAN → 4970105191919435
    - Date exp. 21/25
    - CVV 123
- Le 3DS s'affiche et nous le refusons

Après cela ce message nous est retourné sans déclenchement d'évènements
```
{"type":"https:\/\/tools.ietf.org\/html\/rfc2616#section-10","title":"An error occurred","status":500,"detail":"Internal Server Error"}
```

**Reply by francoistomasi** [reply]
Bonjour, avez-vous des nouvelles de ce bug ?

**Reply by AdrienBr** [reply]
Bonjour @francoistomasi , un correctif est passé récemment et pourrait correspondre à votre problème. Pourriez-vous réessayer ?

Attention, la carte de test 4970105191919435 simule un paiement "frictionless", donc vous ne devriez pas avoir de 3DS affiché pour cette carte.

Adrien

**Reply by francoistomasi** [reply]
Bonjour Adrien, merci pour votre retour.

Le 3DS n'est plus demandé pour les cartes `frictionless` mais il n'est plus demandé non plus pour les cartes `3DSV2`, est-ce normal ?

**Reply by AdrienBr** [reply, marked as answer]
Bonjour,
Je viens de tester, le 3DS est bien déclenché, par exemple sur cette carte : 4970105715165150

Bonne journée,
